analyzers/src/SonarAnalyzer.Common/Rules/ShouldImplementExportedInterfacesBase.cs

```diff
@@ -27,86 +27,109 @@
 
 namespace SonarAnalyzer.Rules.Common
 {
-    public abstract class ShouldImplementExportedInterfacesBase : SonarDiagnosticAnalyzer
+    public abstract class ShouldImplementExportedInterfacesBase<TArgumentSyntax, TAttributeSyntax, TSyntaxKind> : SonarDiagnosticAnalyzer
+        where TArgumentSyntax : SyntaxNode
+        where TAttributeSyntax : SyntaxNode
+        where TSyntaxKind : struct
     {
         internal const string DiagnosticId = "S4159";
-        protected const string MessageFormat = "{0} '{1}' on '{2}' or remove this export attribute.";
-        protected const string ActionForInterface = "Implement";
-        protected const string ActionForClass = "Derive from";
+        private const string MessageFormat = "{0} '{1}' on '{2}' or remove this export attribute.";
+        private const string ActionForInterface = "Implement";
+        private const string ActionForClass = "Derive from";
 
-        internal static readonly ImmutableArray<KnownType> ExportAttributes =
+        private readonly DiagnosticDescriptor rule;
+        private readonly ImmutableArray<KnownType> exportAttributes =
             ImmutableArray.Create(
                 KnownType.System_ComponentModel_Composition_ExportAttribute,
-                KnownType.System_ComponentModel_Composition_InheritedExportAttribute
-            );
-    }
+                KnownType.System_ComponentModel_Composition_InheritedExportAttribute);
 
-    public abstract class ShouldImplementExportedInterfacesBase<TArgumentSyntax, TExpressionSyntax, TClassSyntax>
-        : ShouldImplementExportedInterfacesBase
-        where TArgumentSyntax : SyntaxNode
-        where TExpressionSyntax : SyntaxNode
-        where TClassSyntax : SyntaxNode
-    {
-        protected static bool IsOfExportType(ITypeSymbol type, INamedTypeSymbol exportedType) =>
+        protected abstract TSyntaxKind[] SyntaxKinds { get; }
+        protected abstract SeparatedSyntaxList<TArgumentSyntax>? GetAttributeArguments(TAttributeSyntax attributeSyntax);
+        protected abstract SyntaxNode GetAttributeName(TAttributeSyntax attributeSyntax);
+        protected abstract ILanguageFacade<TSyntaxKind> Language { get; }
+        protected abstract bool IsClassOrRecordSyntax(SyntaxNode syntaxNode);
+        protected abstract string GetIdentifier(TArgumentSyntax argumentSyntax);
+        protected abstract SyntaxNode GetExpression(TArgumentSyntax argumentSyntax);
```

Comment:
Please take a look at the methods in SyntaxFacade (Language.Syntax). There is a chance some of them could be extended and that way there will be no need for these abstract methods like GetExpression.

Reply:
GetIdentifier and GetExpression were eliminated by introducing common code to handle them in the future for corresponding facades